### ISSUE TITLE
Internal error has occurred when executing Firebase ML custom model 

### ISSUE THREAD

**Issue by abdou31** [firebase/quickstart-android#913]
### Step 1: Describe your environment

  * Android device: Samsung A3 2016
  * Android OS version: 7.0
  * Google Play Services version: N/A
  * Firebase/Play Services SDK version: 15.0
  
### Step 2: Describe the problem:
I'm trying to using my own custom tflite model with TensorFlow and ml kit, but I got this problem
  
NB: inputs changed every click on the button 
#### Observed Results:
> inputs: [[[F@....

> Internal error has occurred when executing Firebase ML custom model 

#### Relevant Code:

  ```
package com.example.irisdetection;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.graphics.drawable.BitmapDrawable;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.camerakit.CameraKitView;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.FirebaseApp;
import com.google.firebase.ml.common.FirebaseMLException;
import com.google.firebase.ml.common.modeldownload.FirebaseLocalModel;
import com.google.firebase.ml.common.modeldownload.FirebaseModelDownloadConditions;
import com.google.firebase.ml.common.modeldownload.FirebaseModelManager;
import com.google.firebase.ml.common.modeldownload.FirebaseRemoteModel;
import com.google.firebase.ml.custom.FirebaseModelDataType;
import com.google.firebase.ml.custom.FirebaseModelInputOutputOptions;
import com.google.firebase.ml.custom.FirebaseModelInputs;
import com.google.firebase.ml.custom.FirebaseModelInterpreter;
import com.google.firebase.ml.custom.FirebaseModelOptions;
import com.google.firebase.ml.custom.FirebaseModelOutputs;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;

public class MainActivity extends AppCompatActivity {
   // private CameraKitView cameraKitView;
    ImageView btn_open,btn_stop;
    ImageView image;
    File savedphoto;
    FirebaseModelInterpreter firebaseModelInterpreter;
    private BitmapFactory.Options o;
    private BitmapFactory.Options o2;
    private FileInputStream fis;
    ImageView image2;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        btn_open= findViewById(R.id.btn_open);
        image2= findViewById(R.id.imageView);
        //btn_stop= findViewById(R.id.btn_stop);
      //  cameraKitView= findViewById(R.id.camera);
       //cameraKitView.onStart();
       btn_open.setOnClickListener(new View.OnClickListener() {
           @Override
           public void onClick(View v) {
               /* cameraKitView.captureImage(new CameraKitView.ImageCallback() {
                   @Override
                   public void onImage(CameraKitView cameraKitView, final byte[] photo) {
                       savedphoto=new File(Environment.getExternalStorageDirectory(),"photo.jpg");
                       try {
                           cameraKitView.onStart();
                           FileOutputStream outputStream = new FileOutputStream(savedphoto.getPath());
                           outputStream.write(photo);
                           outputStream.close();*/
                           configureLocalModelSource();
                           try {
                               firebaseModelInterpreter = createInterpreter();
                               Bitmap btmp=getYourInputImage();
                               runInference(btmp);

                           }
                           catch (FirebaseMLException ex){
                               Log.e("Erreur",ex.getMessage());
                               Toast.makeText(MainActivity.this, ""+ex.getMessage(), Toast.LENGTH_SHORT).show();
                           }
                           float [][][][] inputs=bitmapToInputArray(getYourInputImage());
                           Toast.makeText(MainActivity.this, "inputs: "+inputs, Toast.LENGTH_SHORT).show();
/*
                       }
                       catch (IOException ex){
                           ex.printStackTrace();
                           Log.e("CKDemo", "Exception in photo callback");

                       }

                   }
               });*/
           }
       });
      /* btn_stop.setOnClickListener(new View.OnClickListener() {
           @Override
           public void onClick(View v) {
               cameraKitView.onStop();
           }
       });*/




    }

        private void configureHostedModelSource() {
            // [START mlkit_cloud_model_source]
            FirebaseModelDownloadConditions.Builder conditionsBuilder =
                    new FirebaseModelDownloadConditions.Builder().requireWifi();
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                // Enable advanced conditions on Android Nougat and newer.
                conditionsBuilder = conditionsBuilder
                        .requireCharging()
                        .requireDeviceIdle();
            }
            FirebaseModelDownloadConditions conditions = conditionsBuilder.build();

            // Build a remote model source object by specifying the name you assigned the model
            // when you uploaded it in the Firebase console.
            FirebaseRemoteModel cloudSource = new FirebaseRemoteModel.Builder("iristf")
                    .enableModelUpdates(true)
                    .setInitialDownloadConditions(conditions)
                    .setUpdatesDownloadConditions(conditions)
                    .build();
            FirebaseModelManager.getInstance().registerRemoteModel(cloudSource);
            // [END mlkit_cloud_model_source]
        }

        private void configureLocalModelSource() {
            // [START mlkit_local_model_source]
            FirebaseLocalModel localSource =
                    new FirebaseLocalModel.Builder("iristf")  // Assign a name to this model
                            .setAssetFilePath("iristf.tflite")
                            .build();
            FirebaseModelManager.getInstance().registerLocalModel(localSource);
            // [END mlkit_local_model_source]
        }

        private FirebaseModelInterpreter createInterpreter() throws FirebaseMLException {
            // [START mlkit_create_interpreter]
            FirebaseModelOptions options = new FirebaseModelOptions.Builder()
                   // .setRemoteModelName("iristf")
                    .setLocalModelName("iristf")
                    .build();
            FirebaseModelInterpreter firebaseInterpreter =
                    FirebaseModelInterpreter.getInstance(options);
            // [END mlkit_create_interpreter]

            return firebaseInterpreter;
        }

        private FirebaseModelInputOutputOptions createInputOutputOptions() throws FirebaseMLException {
            // [START mlkit_create_io_options]
            FirebaseModelInputOutputOptions inputOutputOptions =
                    new FirebaseModelInputOutputOptions.Builder()
                            .setInputFormat(0, FirebaseModelDataType.FLOAT32, new int[]{1, 112, 112, 3})
                            .setOutputFormat(0, FirebaseModelDataType.FLOAT32, new int[]{1, 80})
                            .build();
            // [END mlkit_create_io_options]

            return inputOutputOptions;
        }

        private float[][][][] bitmapToInputArray(Bitmap bitmap) {
            // [START mlkit_bitmap_input]

            bitmap = Bitmap.createScaledBitmap(bitmap, 112, 112, true);

            int batchNum = 0;
            float[][][][] input = new float[1][112][112][3];
            for (int x = 0; x < 112; x++) {
                for (int y = 0; y < 112; y++) {
                    int pixel = bitmap.getPixel(x, y);
                    // Normalize channel values to [-1.0, 1.0]. This requirement varies by
                    // model. For example, some models might require values to be normalized
                    // to the range [0.0, 1.0] instead.
                    input[batchNum][x][y][0] = (Color.red(pixel) - 127) / 128.0f;
                    input[batchNum][x][y][1] = (Color.green(pixel) - 127) / 128.0f;
                    input[batchNum][x][y][2] = (Color.blue(pixel) - 127) / 128.0f;
                }
            }
            // [END mlkit_bitmap_input]

            return input;
        }

        private void runInference(Bitmap bitmap2) throws FirebaseMLException {
            FirebaseModelInterpreter firebaseInterpreter = createInterpreter();
            bitmap2=getYourInputImage();
            float[][][][] input = bitmapToInputArray(bitmap2);
            FirebaseModelInputOutputOptions inputOutputOptions = createInputOutputOptions();

            // [START mlkit_run_inference]
            FirebaseModelInputs inputs = new FirebaseModelInputs.Builder()
                    .add(input)  // add() as many input arrays as your model requires
                    .build();
            firebaseInterpreter.run(inputs, inputOutputOptions)
                    .addOnSuccessListener(
                            new OnSuccessListener<FirebaseModelOutputs>() {
                                @Override
                                public void onSuccess(FirebaseModelOutputs result) {
                                    // [START_EXCLUDE]
                                    // [START mlkit_read_result]
                                    float[][] output = result.getOutput(40);
                                    float[] probabilities = output[40];
                                    Toast.makeText(MainActivity.this, "probalitities: "+probabilities+"  output: "+output, Toast.LENGTH_SHORT).show();
                                    // [END mlkit_read_result]
                                    // [END_EXCLUDE]
                                }
                            })
                    .addOnFailureListener(
                            new OnFailureListener() {
                                @Override
                                public void onFailure(@NonNull Exception e) {
                                    Log.e("Error",e.getMessage());
                                    Toast.makeText(MainActivity.this, ""+e.getMessage(), Toast.LENGTH_SHORT).show();
                                    // Task failed with an exception
                                    // ...
                                }
                            });
            // [END mlkit_run_inference]
        }

        private void useInferenceResult(float[] probabilities) throws IOException {
            // [START mlkit_use_inference_result]
           /* BufferedReader reader = new BufferedReader(
                    new InputStreamReader(getAssets().open("retrained_labels.txt")));*/
            for (int i = 0; i < probabilities.length; i++) {
             //   String label = reader.readLine();
                Log.i("MLKit", String.format("%1.4f", probabilities[i]));
            }
            // [END mlkit_use_inference_result]
        }

        private Bitmap getYourInputImage() {
            // This method is just for show
            BitmapDrawable drawable = (BitmapDrawable) image2.getDrawable();
            Bitmap bitmap = drawable.getBitmap();
            Bitmap bitmap2=Bitmap.createScaledBitmap(bitmap,112,112,true);
            return bitmap2;
          //  return Bitmap.createBitmap(112, 112, Bitmap.Config.ALPHA_8);
        }

    public Bitmap decodeFile(File f) {
        Bitmap b = null;
        try {
            // Decode image size
            o = new BitmapFactory.Options();
            o.inJustDecodeBounds = true;

            fis = new FileInputStream(f);
            BitmapFactory.decodeStream(fis, null, o);
            fis.close();
            int IMAGE_MAX_SIZE = 1000;
            int scale = 1;
            if (o.outHeight > IMAGE_MAX_SIZE || o.outWidth > IMAGE_MAX_SIZE) {
                scale = (int) Math.pow(
                        2,
                        (int) Math.round(Math.log(IMAGE_MAX_SIZE
                                / (double) Math.max(o.outHeight, o.outWidth))
                                / Math.log(0.5)));
            }

            // Decode with inSampleSize
            o2 = new BitmapFactory.Options();
            o2.inSampleSize = scale;
            fis = new FileInputStream(f);
            b = BitmapFactory.decodeStream(fis, null, o2);
            fis.close();

        } catch (IOException e) {
            e.printStackTrace();
        }

        return b;
    }

  /*  @Override
    protected void onStart() {
        super.onStart();
        cameraKitView.onStart();
    }
    @Override
    protected void onResume() {
        super.onResume();
        cameraKitView.onResume();
    }
    @Override
    protected void onPause() {
        cameraKitView.onPause();
        super.onPause();
    }
    @Override
    protected void onStop() {
        cameraKitView.onStop();
        super.onStop();
    }
    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        cameraKitView.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }*/

    }


  ```

How can I solve this problem?

**Comment by samtstern** [comment]
@abdou31 this doesn't sound like it's an issue with the quickstart and the code presented is far too big for us to hunt for the error in this forum. Since it sounds like you need general help debugging, I would suggest posting on StackOverflow or a similar forum.

**Comment by abdou31** [comment]
I really need help.
For now, I can't post anymore on StackOverflow for some reasons.
If you would really help me, I will reduce the source code and give just that it related to the error.

**Comment by cxx120** [comment]
From a quick glance of your code, looks like there is something wrong with how your access the output?

"float[][] output = result.getOutput(40);"

Would this be result.getOutput(0),  you specified
...
.setOutputFormat(0, FirebaseModelDataType.FLOAT32, new int[]{1, 80})

in the InputOutputOptions?

**Comment by abdou31** [comment]
I corrected some mistakes and I have changed result.getouput(40) by 0.
I get some results with Toast ( landmarks of eye region of a imageview ).
For the socond line , that was correct because I have setting units to 80 in the logits layer ( ouptut layer )
80 is meaning ( 40 *2), 40 the number of landmarks created on the eye regions.
(x,y)=> 80.
x=>40.
y=>40.
